CODESTYLE.md

```diff
@@ -0,0 +1,208 @@
+
+[Testify]([https://github.com/stretchr/testify) "A toolkit with common assertions and mocks that plays nicely with the standard library"
+
+[Mockery](https://github.com/vektra/mockery) "A mock code autogenerator for golang"
+
+Mocks should go into a subpackage of the interface that implements
+to generate a Mock for a interface run:
+
+```go
+mockery -name InterfaceName
+```
+
+### Test fixtures
+
+We use Golden Files trick, to keep our test fixtures up to date
+
+All files should have a suffix `.golden` and update flag for tests should be `--update`
+
+example:
+
+```go
+var update = flag.Bool("update", false, "update .golden files")
+func TestSomething(t *testing.T) {
+  actual := doSomething()
+  golden := filepath.Join("testdata", tc.Name+"".golden"")
+  if *update {
+    ioutil.WriteFile(golden, actual, 0644)
+  }
+  expected, _ := ioutil.ReadFile(golden)
+  assert.Equal(t, actual, expected)
+}
+```
+
+
+### E2E
+
+[Godog](https://github.com/DATA-DOG/godog) "Cucumber for golang"
+
+
+## General
+
+### Valid Objects
+ 
+Create valid objects at construction time, use several constructors to have variants of the object, but all should be valid.
+
+Valid here means, that further call should be needed to use the object.
+
+Don't do this
+
+```go
+email := NewEMail()
+email.SetTo("kcoin@kowala.tech")
+``` 
+the developer might not me aware that it has to set de TO before using EMail
+
+If you require and IP why not
+
+```go
+email := NewMailer("kcoin@kowala.tech")
+```
+
+### Avoid too many arguments
+
+"Functions should have a small number of arguments. No argument is best, followed by one, two, and three. More than three is very questionable and should be avoided with prejudice."
+(http://www.informit.com/articles/article.aspx?p=1375308)
+
+Consider using struct of values, optional params or a builder.
+
+Don't do this
+
+```go
+postLetter(firstName string, lastName string, street string, city string, postcode string, flatNumber int)
+```
+
+why not
+
+```go
+postLetter(personName PersonName, address Address)
+```
+
+### Readability over "smart" code
+
+Write code for humans first, try to express the intent with function and variable names.
+
+From 
+```go
+currentBlock := val.chain.CurrentBlock()
+if currentBlock.Number().Cmp(big.NewInt(0)) == 0 {
+    return
+}
+```
+
+To 
+```go
+currentBlock := val.chain.CurrentBlock()
+if isFirstBlock(currentBlock) {
+    return
+}
+
+func isFirstBlock(block Block) bool {
+    return block.Number().Cmp(big.NewInt(0)) == 0
+}
+```
+### Named returns
+
+Go supports named returns, but they are discouraged from standard go style. They should be used on very small functions only.
+Some are well known problems like shadowing. 
+
+```go
+func ReadFull(r Reader, buf []byte) (n int, err error) {
+    for len(buf) > 0 && err == nil {
+        var nr int
+        nr, err = r.Read(buf)
+        n += nr
+        buf = buf[nr:]
+    }
+    return
+}
+```
+
+### Interfaces
+
+To ease testing consider using interfaces  
```

Comment:
New concepts should always be expressed in interfaces first. In the ethereum code that's an issue, but it all our other code we can do it.